### DISCUSSION TITLE
confused about workspace topologies and "out-of-tree" applications

### DISCUSSION THREAD

**Discussion by rjmaris** [zephyrproject-rtos/zephyr#58895] [General]
Some time ago, I started to create a copy of a sample project according to the T2-style workspace topology, where I followed the instructions [here](docs.zephyrproject.org/latest/develop/west/workspaces.html). This allows me to build modified sample code built for boards that are inside the `zephyr/boards` directory - regardless git commit checkout.
This not only allows me to maintain code in a private repository which can be simply restored when Zephyrproject would be reinstalled or installed on another computer.
Just now I found the out-of-tree board sample, including local boards directory.

This confuses me: what's the conceptual difference between workspace T2 topology application and "freestanding application"? I.e. what's the advantage in terms of e.g. complexity between the two options? This question is independent of whether a custom board is assigned or not.

**Comment by yashi** [comment]
I don't really use the "freestanding application" so I might be wrong but

- T2: you want to manage your app along with Zephyr and other modules.  You want it to be buildable, say, after two years or 3 Zephyr releases.  It is possible because your app's repo has the hash values for each modules and you can reproduce it any time.

- freestanding: It can be built against a version of Zephyr and versions of other modules.  But may not be in near future.  But you can document which version it requires.  It can target multiple versions of Zephyr or some combination of Zephyr and modules if you need.

**Reply by rjmaris** [reply]
What you write now clarifies it a little bit for me. I'd now clearly prefer the T2 approach, but still see not really a difference. Well e.g. the usage of a `west.yml` in a T2 topology, as is shown in the existing `example-application` (which I downloaded some time ago). The `out_of_tree_board`, copied over from `samples/application_development` cointains a (optionally) yaml file.

So I still do not have a clear view of the two, except that **workspace** is something else then plain **out of tree**. But strictly seen, T2 topology *is* out of tree. Or even more T3 when I acknowledge that `west.yml` is not part of the app directories...
